File: sdk/go/kafka/kafka_consumer.go
// Package kafka is the real Kafka consumer adapter for the BYO worker SDK,
// implementing the broker-agnostic client.Consumer SPI on top of
// github.com/segmentio/kafka-go.
//
// It lives in a NESTED module (see go.mod) so the root batch-worker-sdk-go
// module stays zero-dependency: the SDK core only knows the client.Consumer
// interface and never imports a Kafka library. Tenants who want real Kafka
// pull in this module; tenants who embed the SDK and bring their own broker
// glue do not.
//
// Wire behavior implemented here mirrors byo-sdk-guide.md §1.2 / §1.5 / §1.9:
//   - wildcard topic subscription batch.task.dispatch.<tenant>.*
//   - consumer group g-sdk-<tenantId>-<workerCode>
//   - SASL/SCRAM-SHA-512 + TLS support (config-driven; PLAINTEXT when no creds)
//   - JSON deserialize -> client.MessagePipeline (schemaVersion reject,
//     tenant self-check, capacity backpressure)
//   - MANUAL offset commit: commit ONLY after the pipeline accepts; never on
//     reject (unknown schemaVersion) or foreign-tenant drop -> redelivery
//   - partition pause/resume modeled by stopping/resuming the fetch loop
package kafka

import (
	"context"
	"crypto/tls"
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/pinpols/file-batch-system/batch-worker-sdk-go/client"
	kgo "github.com/segmentio/kafka-go"
	"github.com/segmentio/kafka-go/sasl/scram"
)

// DefaultTopicRefreshInterval is how often the adapter re-lists broker topics to
// pick up newly created per-task-type dispatch topics
// (batch.task.dispatch.<tenant>.<taskType>). kafka-go's GroupReader takes a
// fixed GroupTopics slice at construction and does NOT support regex/wildcard
// subscriptions, so wildcard is emulated by periodic metadata polling +
// prefix-filtering + reader restart when the matched set changes.
const DefaultTopicRefreshInterval = 5 * time.Minute

// DefaultFetchMaxWait bounds how long a single underlying FetchMessage blocks
// before returning so Poll() stays responsive to Wakeup()/Pause().
const DefaultFetchMaxWait = 1 * time.Second

// Config drives the Kafka adapter. Credentials come from here (sourced by the
// caller from env / secret store) and are NEVER read from message payloads
// (byo-sdk-guide §1.2/§1.8).
type Config struct {
	// Brokers is the bootstrap broker list, e.g. ["localhost:19092"].
	Brokers []string
	// TenantID scopes the wildcard topic prefix and the consumer group, and is
	// the value the pipeline tenant self-check (§1.9) compares against.
	TenantID string
	// WorkerCode identifies this worker within the tenant; combined with
	// TenantID into the consumer group g-sdk-<tenantId>-<workerCode>.
	WorkerCode string

	// SASLUsername / SASLPassword enable SASL/SCRAM-SHA-512 when both are set.
	// When empty, the adapter connects over PLAINTEXT (local integration).
	SASLUsername string
	SASLPassword string

	// TLSConfig, when non-nil, wraps the connection in TLS. SASL/SCRAM in
	// production is normally paired with TLS; locally we use neither.
	TLSConfig *tls.Config

	// TopicRefreshInterval overrides DefaultTopicRefreshInterval when > 0.
	TopicRefreshInterval time.Duration
	// FetchMaxWait overrides DefaultFetchMaxWait when > 0.
	FetchMaxWait time.Duration

	// Logger is optional; defaults to log.Default().
	Logger *log.Logger
}

// topicPrefix returns the wildcard prefix batch.task.dispatch.<tenant>. that a
// topic must start with to belong to this tenant.
func (c Config) topicPrefix() string {
	return "batch.task.dispatch." + c.TenantID + "."
}

// groupID returns the per-worker consumer group g-sdk-<tenantId>-<workerCode>.
func (c Config) groupID() string {
	return "g-sdk-" + c.TenantID + "-" + c.WorkerCode
}

// Consumer is the kafka-go-backed implementation of client.Consumer.
//
// Pause/Resume have no native equivalent on kafka-go's GroupReader, so they are
// modeled by gating the fetch loop: while paused, Poll() blocks (parking until
// resumed or woken) instead of fetching, which stops new records from flowing —
// equivalent in effect to assignment.pause(). Offsets already fetched are not
// committed until Commit() is called after pipeline acceptance.
type Consumer struct {
	cfg    Config
	logger *log.Logger
	dialer *kgo.Dialer
	client *kgo.Client

	mu     sync.Mutex
	reader *kgo.Reader
	topics []string
	paused bool
	woken  bool

	// pending holds messages fetched-but-not-yet-committed, so Commit() can
	// acknowledge exactly the offsets the pipeline accepted.
	pending []kgo.Message

	resumeCh chan struct{} // signaled on Resume()/Wakeup() to unpark a paused Poll
	ctx      context.Context
	cancel   context.CancelFunc

	lastRefresh time.Time
}

// compile-time assertion that *Consumer satisfies the phase-2 SPI.
var _ client.Consumer = (*Consumer)(nil)

// NewConsumer builds the adapter, performs an initial topic discovery, and
// constructs the GroupReader. It returns an error if no matching topic exists
// yet or the broker is unreachable.
func NewConsumer(cfg Config) (*Consumer, error) {
	if len(cfg.Brokers) == 0 {
		return nil, fmt.Errorf("kafka: no brokers configured")
	}
	if cfg.TenantID == "" || cfg.WorkerCode == "" {
		return nil, fmt.Errorf("kafka: tenantID and workerCode are required")
	}
	logger := cfg.Logger
	if logger == nil {
		logger = log.Default()
	}

	dialer, err := buildDialer(cfg)
	if err != nil {
		return nil, err
	}

	transport, err := buildTransport(cfg)
	if err != nil {
		return nil, err
	}
	kc := &kgo.Client{
		Addr:      kgo.TCP(cfg.Brokers...),
		Transport: transport,
	}

	ctx, cancel := context.WithCancel(context.Background())
	c := &Consumer{
		cfg:      cfg,
		logger:   logger,
		dialer:   dialer,
		client:   kc,
		resumeCh: make(chan struct{}, 1),
		ctx:      ctx,
		cancel:   cancel,
	}

	topics, err := c.discoverTopics()
	if err != nil {
		cancel()
		return nil, err
	}
	if len(topics) == 0 {
		cancel()
		return nil, fmt.Errorf("kafka: no topics match prefix %q (create the dispatch topic first)", cfg.topicPrefix())
	}
	c.topics = topics
	c.reader = c.newReader(topics)
	c.lastRefresh = time.Now()
	logger.Printf("INFO kafka adapter subscribed group=%s topics=%v", cfg.groupID(), topics)
	return c, nil
}

// buildDialer returns a kafka-go Dialer configured with SASL/SCRAM-SHA-512 and
// TLS when credentials are supplied, or a plain dialer otherwise.
func buildDialer(cfg Config) (*kgo.Dialer, error) {
	d := &kgo.Dialer{
		Timeout:   10 * time.Second,
		DualStack: true,
	}
	if cfg.SASLUsername != "" && cfg.SASLPassword != "" {
		mech, err := scram.Mechanism(scram.SHA512, cfg.SASLUsername, cfg.SASLPassword)
		if err != nil {
			return nil, fmt.Errorf("kafka: build SCRAM-SHA-512 mechanism: %w", err)
		}
		d.SASLMechanism = mech
	}
	if cfg.TLSConfig != nil {
		d.TLS = cfg.TLSConfig
	}
	return d, nil
}

// buildTransport mirrors the dialer's auth settings for the metadata client
// (kgo.Client uses a Transport rather than a Dialer).
func buildTransport(cfg Config) (*kgo.Transport, error) {
	t := &kgo.Transport{}
	if cfg.SASLUsername != "" && cfg.SASLPassword != "" {
		mech, err := scram.Mechanism(scram.SHA512, cfg.SASLUsername, cfg.SASLPassword)
		if err != nil {
			return nil, fmt.Errorf("kafka: build SCRAM-SHA-512 transport mechanism: %w", err)
		}
		t.SASL = mech
	}
	if cfg.TLSConfig != nil {
		t.TLS = cfg.TLSConfig
	}
	return t, nil
}

// discoverTopics lists all broker topics and returns those matching the
// tenant's wildcard prefix. This is the wildcard-subscription workaround:
// kafka-go does NOT support regex topics on a GroupReader, so we resolve the
// concrete topic set from cluster metadata.
func (c *Consumer) discoverTopics() ([]string, error) {
	ctx, cancel := context.WithTimeout(c.ctx, 10*time.Second)
	defer cancel()
	resp, err := c.client.Metadata(ctx, &kgo.MetadataRequest{})
	if err != nil {
		return nil, fmt.Errorf("kafka: metadata request: %w", err)
	}
	prefix := c.cfg.topicPrefix()
	var matched []string
	for _, t := range resp.Topics {
		if t.Error != nil {
			continue
		}
		if strings.HasPrefix(t.Name, prefix) {
			matched = append(matched, t.Name)
		}
	}
	return matched, nil
}

// newReader constructs a GroupReader over the given concrete topic set with
// MANUAL commit (CommitInterval: 0) so offsets are only acknowledged via
// explicit CommitMessages after the pipeline accepts.
func (c *Consumer) newReader(topics []string) *kgo.Reader {
	maxWait := c.cfg.FetchMaxWait
	if maxWait <= 0 {
		maxWait = DefaultFetchMaxWait
	}
	return kgo.NewReader(kgo.ReaderConfig{
		Brokers:        c.cfg.Brokers,
		GroupID:        c.cfg.groupID(),
		GroupTopics:    topics,
		Dialer:         c.dialer,
		CommitInterval: 0, // MANUAL commit
		MaxWait:        maxWait,
	})
}

// maybeRefreshTopics periodically re-lists topics; if the matched set changed it
// rebuilds the reader so newly created <taskType> topics get consumed. Any
// uncommitted pending messages are dropped on rebuild (they will be redelivered
// since they were never committed), preserving at-least-once.
//
// It is called with c.mu held. The old reader is NOT closed here — closing a
// kafka-go reader can block (it leaves the group), and doing so under c.mu would
// stall a concurrent Wakeup()/Close()/Commit(). Instead the displaced reader is
// returned so the caller can Close() it AFTER releasing the lock.
func (c *Consumer) maybeRefreshTopics() (old *kgo.Reader) {
	interval := c.cfg.TopicRefreshInterval
	if interval <= 0 {
		interval = DefaultTopicRefreshInterval
	}
	if time.Since(c.lastRefresh) < interval {
		return nil
	}
	c.lastRefresh = time.Now()
	topics, err := c.discoverTopics()
	if err != nil {
		c.logger.Printf("WARN kafka topic refresh failed: %v", err)
		return nil
	}
	if len(topics) == 0 || sameTopics(topics, c.topics) {
		return nil
	}
	c.logger.Printf("INFO kafka topic set changed old=%v new=%v; rebuilding reader", c.topics, topics)
	old = c.reader
	c.reader = c.newReader(topics)
	c.topics = topics
	c.pending = nil
	return old
}

// sameTopics compares two topic sets ignoring order.
func sameTopics(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	seen := make(map[string]struct{}, len(a))
	for _, t := range a {
		seen[t] = struct{}{}
	}
	for _, t := range b {
		if _, ok := seen[t]; !ok {
			return false
		}
	}
	return true
}

// Poll fetches one record (kafka-go has no batch fetch; we return single-record
// batches) and returns it as a client.Record. It returns (nil, false) when the
// consumer has been woken/closed. While paused, it parks until Resume/Wakeup
// rather than fetching — the partition-pause emulation.
func (c *Consumer) Poll() ([]client.Record, bool) {
	for {
		c.mu.Lock()
		if c.woken {
			c.mu.Unlock()
			return nil, false
		}
		if c.paused {
			c.mu.Unlock()
			// Park until resumed or woken; do not fetch new records.
			select {
			case <-c.resumeCh:
				continue
			case <-c.ctx.Done():
				return nil, false
			}
		}
		oldReader := c.maybeRefreshTopics()
		reader := c.reader
		c.mu.Unlock()

		// Close the displaced reader outside the lock (it can block leaving the
		// group); a nil-check keeps the common no-refresh path allocation-free.
		if oldReader != nil {
			_ = oldReader.Close()
		}

		fetchCtx, cancel := context.WithTimeout(c.ctx, c.fetchTimeout())
		msg, err := reader.FetchMessage(fetchCtx)
		cancel()
		if err != nil {
			c.mu.Lock()
			woken := c.woken
			c.mu.Unlock()
			if woken || c.ctx.Err() != nil {
				return nil, false
			}
			// Timeout / transient: return an empty batch with ok=true so the
			// consume loop can re-check pause/wakeup state and poll again.
			return nil, true
		}

		c.mu.Lock()
		c.pending = append(c.pending, msg)
		c.mu.Unlock()

		return []client.Record{{
			Topic: msg.Topic,
			Key:   string(msg.Key),
			Value: msg.Value,
		}}, true
	}
}

func (c *Consumer) fetchTimeout() time.Duration {
	w := c.cfg.FetchMaxWait
	if w <= 0 {
		w = DefaultFetchMaxWait
	}
	// Give the fetch a little headroom beyond MaxWait.
	return w + 500*time.Millisecond
}

// Pause stops the fetch loop (no new records). In-flight commits still work.
func (c *Consumer) Pause() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.paused = true
}

// Resume unparks a paused Poll and lets fetching continue.
func (c *Consumer) Resume() {
	c.mu.Lock()
	c.paused = false
	c.mu.Unlock()
	c.signalResume()
}

// Commit acknowledges all pending (fetched-and-accepted) offsets. The live
// worker calls Commit() ONLY after the pipeline accepts; on reject / foreign
// tenant the caller must NOT call Commit for that record, so it is redelivered.
//
// Because Poll returns one record per call and the worker commits per accepted
// record, pending normally holds exactly the message just accepted. Any pending
// message left uncommitted at the next rebuild/close is redelivered.
func (c *Consumer) Commit() {
	c.mu.Lock()
	if len(c.pending) == 0 {
		c.mu.Unlock()
		return
	}
	msgs := c.pending
	c.pending = nil
	reader := c.reader
	c.mu.Unlock()

	ctx, cancel := context.WithTimeout(c.ctx, 10*time.Second)
	defer cancel()
	if err := reader.CommitMessages(ctx, msgs...); err != nil {
		c.logger.Printf("ERROR kafka commit offsets failed (will redeliver): %v", err)
		// Put them back so a later Commit can retry.
		c.mu.Lock()
		c.pending = append(msgs, c.pending...)
		c.mu.Unlock()
	}
}

// DropPending discards the not-yet-committed messages without committing, so the
// broker redelivers them. The worker calls this after a reject / foreign-tenant
// drop instead of Commit(), keeping the offset withheld (§1.9 / §A).
func (c *Consumer) DropPending() {
	c.mu.Lock()
	c.pending = nil
	c.mu.Unlock()
}

// Wakeup interrupts an in-progress Poll (graceful stop, §1.6). Idempotent and
// goroutine-safe.
func (c *Consumer) Wakeup() {
	c.mu.Lock()
	c.woken = true
	c.mu.Unlock()
	c.signalResume() // unpark a parked-paused Poll
	c.cancel()       // interrupt an in-flight FetchMessage
}

// Close releases the reader and metadata client.
func (c *Consumer) Close() {
	c.mu.Lock()
	reader := c.reader
	c.reader = nil
	c.mu.Unlock()
	c.cancel()
	if reader != nil {
		_ = reader.Close()
	}
}

// signalResume does a non-blocking send on resumeCh to wake a parked Poll.
func (c *Consumer) signalResume() {
	select {
	case c.resumeCh <- struct{}{}:
	default:
	}
}

File: sdk/typescript/src/kafka/kafkaConsumer.ts
/**
 * Real kafkajs adapter for the BYO worker SDK (byo-sdk-guide §1.2 / §1.9,
 * wire-protocol §2.1 / §A).
 *
 * This is the ONLY module (besides its integration test) that imports kafkajs.
 * The core `src/*` stays zero-dependency: this adapter IMPLEMENTS the phase-2
 * `Consumer` interface and forwards each record through a `MessagePipeline`.
 *
 * Responsibilities (the platform contract the core decision rules describe):
 *   - wildcard subscribe `batch.task.dispatch.<tenant>.*` via a RegExp topic
 *   - consumer group `g-sdk-<tenantId>-<workerCode>`
 *   - SASL/SCRAM-SHA-512 + SSL config support (PLAINTEXT when no sasl given)
 *   - JSON UTF-8 deserialize, run through MessagePipeline (schema reject /
 *     tenant self-check §1.9 / backpressure)
 *   - manual offset commit (autoCommit:false): commit ONLY when the pipeline
 *     ACCEPTS, so rejected/dropped messages are redelivered by the platform
 *   - partition pause/resume driven by the backpressure decision
 *   - credentials come from config/env, NEVER from the message payload
 */

import {
  Kafka,
  logLevel as kafkaLogLevel,
  type Consumer as KafkaJsConsumer,
  type EachMessagePayload,
  type SASLOptions,
} from "kafkajs";

import {
  MessagePipeline,
  dispositionOf,
  type Consumer,
  type ConsumerRecord,
  type Logger,
  type MessageDisposition,
  consoleLogger,
} from "../client/consumer.ts";
import {
  offsetAction,
  resolveFromBeginning,
  loweredCeiling,
  commitBlockedByWithheld,
} from "./offsetPolicy.ts";

/** SASL config (only SCRAM-SHA-512 is supported per byo-sdk-guide §1.2). */
export interface KafkaSaslConfig {
  mechanism: "scram-sha-512";
  username: string;
  password: string;
}

/**
 * Connection + identity config. Credentials live here (sourced from env/secret
 * by the caller) — never read from a dispatch payload.
 */
export interface KafkaConsumerConfig {
  /** bootstrap brokers, e.g. ["localhost:19092"]. */
  brokers: string[];
  /** tenant this worker serves; drives the topic regex + group + self-check. */
  tenantId: string;
  /** worker code; part of the consumer group id. */
  workerCode: string;
  /** TLS on the wire (SASL_SSL). false/omitted → plaintext transport. */
  ssl?: boolean;
  /** SASL/SCRAM-SHA-512. Omit for PLAINTEXT (local broker). */
  sasl?: KafkaSaslConfig;
  /** override the generated clientId. */
  clientId?: string;
  /** override the generated consumer group id (mainly for test isolation). */
  groupId?: string;
  /** override the wildcard topic regex (mainly for test isolation). */
  topicRegex?: RegExp;
  /** start from earliest on a brand-new group (default true; handy for tests). */
  fromBeginning?: boolean;
}

/**
 * Build the node-direct subscription regex for
 * `batch.task.dispatch.<workerType>.node.<workerCode>` — the topic the platform
 * dispatches THIS worker's tasks to (base-first, aligned with built-in workers'
 * AbstractTaskConsumer.topicPattern()). The old tenant-first
 * `batch.task.dispatch.<tenant>.*` is never published to; cross-tenant safety is
 * enforced by the per-message tenant self-check (§1.9), not the topic name.
 */
export function dispatchTopicRegex(workerCode: string): RegExp {
  const safe = workerCode.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return new RegExp(`^batch\\.task\\.dispatch\\..+\\.node\\.${safe}$`);
}

/** Consumer group id per byo-sdk-guide §1.2: g-sdk-<tenantId>-<workerCode>. */
export function consumerGroupId(tenantId: string, workerCode: string): string {
  return `g-sdk-${tenantId}-${workerCode}`;
}

/**
 * kafkajs adapter implementing the phase-2 `Consumer` seam.
 *
 * Usage: wire a `MessagePipeline` (from core), then either drive it yourself by
 * passing the pipeline's `onMessage` into `start`, or use `runPipeline` which
 * adds the manual-commit + pause/resume policy on top of the pipeline outcome.
 */
export class KafkaConsumerAdapter implements Consumer {
  readonly #kafka: Kafka;
  readonly #consumer: KafkaJsConsumer;
  readonly #config: KafkaConsumerConfig;
  readonly #logger: Logger;
  readonly #topicRegex: RegExp;
  readonly #groupId: string;
  #running = false;
  #paused = false;
  /** topics actually matched by the wildcard regex (learned at GROUP_JOIN). */
  readonly #assignedTopics = new Set<string>();
  /**
   * Per-partition commit CEILING: the LOWEST offset withheld (rejected schema /
   * foreign tenant / not-for-us) on that partition. No later commit may cross it,
   * so the withheld record is redelivered on the next rebalance/restart. Keyed by
   * `${topic}/${partition}`. This is the kafkajs equivalent of the Go SDK's
   * `withheld` ceiling — it lets the consumer KEEP fetching subsequent records
   * (no head-of-line block) while guaranteeing the withheld offset is never
   * silently skipped.
   */
  readonly #withheld = new Map<string, string>();

  constructor(config: KafkaConsumerConfig, logger: Logger = consoleLogger) {
    this.#config = config;
    this.#logger = logger;
    this.#topicRegex = config.topicRegex ?? dispatchTopicRegex(config.workerCode);
    this.#groupId =
      config.groupId ?? consumerGroupId(config.tenantId, config.workerCode);

    const sasl: SASLOptions | undefined = config.sasl
      ? {
          // kafkajs mechanism name for SCRAM-SHA-512
          mechanism: "scram-sha-512",
          username: config.sasl.username,
          password: config.sasl.password,
        }
      : undefined;

    this.#kafka = new Kafka({
      clientId:
        config.clientId ?? `sdk-${config.tenantId}-${config.workerCode}`,
      brokers: config.brokers,
      // ssl + sasl → SASL_SSL; sasl only → SASL_PLAINTEXT; neither → PLAINTEXT
      ssl: config.ssl ?? false,
      sasl,
      logLevel: kafkaLogLevel.NOTHING,
    });

    this.#consumer = this.#kafka.consumer({ groupId: this.#groupId });

    // learn the concrete topics the wildcard regex matched, so FSM-level
    // pause()/resume() (PAUSED / DRAINING, §1.5) can act on whole topics.
    this.#consumer.on(this.#consumer.events.GROUP_JOIN, (e) => {
      for (const topic of Object.keys(e.payload.memberAssignment ?? {})) {
        this.#assignedTopics.add(topic);
      }
    });
  }

  /** The resolved consumer group id (for assertions / logging). */
  get groupId(): string {
    return this.#groupId;
  }

  /** The resolved wildcard topic regex. */
  get topicRegex(): RegExp {
    return this.#topicRegex;
  }

  /**
   * Connect, subscribe to the wildcard topic, and run with manual commit + the
   * full offset/seek/pause policy.
   *
   * `onMessage` is the production seam (the lifecycle passes
   * `MessagePipeline.onMessage` through it). It returns a {@link MessageDisposition}
   * the adapter acts on AFTER the record is handled:
   *   - "commit"       → advance past this offset (accepted task / poison skip)
   *   - "withhold"     → record commit ceiling + keep consuming; the offset is NEVER crossed (§A/§1.9)
   *   - "backpressure" → seek back + pause; resumed when a slot frees (§1.5/§2)
   *
   * This replaces the old `start()` that only forwarded records and never
   * committed (offset never advanced → every restart replayed the partition).
   */
  async start(
    onMessage: (r: ConsumerRecord) => Promise<MessageDisposition>,
  ): Promise<void> {
    await this.#runLoop(onMessage);
  }

  /**
   * Run the consumer with a full `MessagePipeline`. Convenience wrapper over
   * {@link start}: feeds each record through the pipeline and maps the outcome to
   * the offset disposition via the shared {@link dispositionOf}.
   */
  async runPipeline(pipeline: MessagePipeline): Promise<void> {
    await this.#runLoop(async (record) =>
      dispositionOf(await pipeline.onMessage(record)),
    );
  }

  /** Shared connect → subscribe → run-with-policy loop for start()/runPipeline(). */
  async #runLoop(
    onEach: (r: ConsumerRecord) => Promise<MessageDisposition>,
  ): Promise<void> {
    await this.#consumer.connect();
    await this.#consumer.subscribe({
      topic: this.#topicRegex,
      // default to latest (auto.offset.reset=latest, parity with the other four
      // SDKs). fromBeginning:true would replay the whole topic on a new group.
      fromBeginning: resolveFromBeginning(this.#config.fromBeginning),
    });
    this.#running = true;
    this.#logger.info("kafka consumer subscribed", {
      group: this.#groupId,
      topicRegex: this.#topicRegex.source,
    });

    await this.#consumer.run({
      autoCommit: false,
      eachMessage: async (payload: EachMessagePayload) => {
        this.#assignedTopics.add(payload.topic);
        const value = payload.message.value?.toString("utf8") ?? "";
        const record: ConsumerRecord = {
          value,
          meta: {
            topic: payload.topic,
            partition: payload.partition,
            offset: payload.message.offset,
          },
        };

        const disposition = await onEach(record);
        const key = `${payload.topic}/${payload.partition}`;
        const action = offsetAction(
          disposition,
          payload.topic,
          payload.partition,
          payload.message.offset,
        );

        if (action.type === "commit") {
          // never commit past a record withheld earlier on this partition — that
          // would silently skip it (§A/§1.9). Below the ceiling → advance normally.
          if (commitBlockedByWithheld(this.#withheld.get(key), payload.message.offset)) {
            this.#logger.warn("commit withheld: offset at/past partition ceiling", {
              topic: action.topic,
              partition: action.partition,
              offset: payload.message.offset,
              ceiling: this.#withheld.get(key),
            });
            return;
          }
          // manual commit: advance past this offset (offset + 1 per kafkajs)
          await this.#consumer.commitOffsets([
            { topic: action.topic, partition: action.partition, offset: action.offset },
          ]);
          return;
        }

        if (action.type === "withhold") {
          // record the commit ceiling and KEEP consuming (no pause): the withheld
          // offset is never committed → redelivered later, while subsequent records
          // on this partition still flow (no head-of-line block). Mirrors the Go
          // SDK's `withheld` ceiling; reaches the same invariant as Java's seek-back
          // WITHOUT stalling the partition on a poison/foreign record.
          this.#withheld.set(key, loweredCeiling(this.#withheld.get(key), action.offset));
          this.#logger.warn("offset withheld (ceiling set, still consuming)", {
            topic: action.topic,
            partition: action.partition,
            offset: action.offset,
          });
          return;
        }

        // backpressure: seek BACK to this offset then pause the partition, so (a)
        // this message is redelivered on resume and (b) NO later commit crosses it.
        // Unlike withhold, this pause is TEMPORARY — resumed when a slot frees.
        this.#consumer.seek({
          topic: action.topic,
          partition: action.partition,
          offset: action.offset,
        });
        this.#consumer.pause([
          { topic: action.topic, partitions: [action.partition] },
        ]);
        this.#paused = true;
        this.#logger.warn("seek-back + pause partition (backpressure)", {
          topic: action.topic,
          partition: action.partition,
          offset: action.offset,
          reason: action.reason,
        });
      },
    });
  }

  /** Resume a specific partition (call when a concurrency slot frees up). */
  resumePartition(topic: string, partition: number): void {
    this.#consumer.resume([{ topic, partitions: [partition] }]);
    this.#paused = false;
    this.#logger.info("resumed partition", { topic, partition });
  }

  /** Interrupt the poll loop (byo-sdk-guide §1.6 graceful stop). */
  async wakeup(): Promise<void> {
    this.#running = false;
    await this.#consumer.stop();
  }

  /** Disconnect entirely (cleanup). */
  async disconnect(): Promise<void> {
    this.#running = false;
    await this.#consumer.disconnect();
  }

  /** Pause ALL assigned partitions (FSM PAUSED/DRAINING, §1.5). */
  pause(): void {
    const topics = [...this.#assignedTopics];
    if (topics.length > 0) {
      // topic-only entry pauses every partition of that topic
      this.#consumer.pause(topics.map((topic) => ({ topic })));
    }
    this.#paused = true;
  }

  /** Resume ALL assigned partitions. */
  resume(): void {
    const paused = this.#consumer.paused();
    if (paused.length > 0) {
      this.#consumer.resume(paused);
    }
    this.#paused = false;
  }

  isPaused(): boolean {
    return this.#paused;
  }

  get isRunning(): boolean {
    return this.#running;
  }
}
